Fabric.Authorization.AccessControl/src/app/modules/access-control/member-add/member-add.component.ts

```diff
@@ -1,129 +1,162 @@
-    return this.selectedRoles.filter(r => r.name === role.name).length > 0;
-  }
+  addMemberWithRoles() {
+    const selected: IFabricPrincipalModel = this.principals.find(p => p.selected);
+    if (!selected) {
+      return;
+    }
 
-  getRoles() {
-    this.roleService
-      .getRolesBySecurableItemAndGrain(
-        this.configService.grain,
-        this.configService.securableItem
-      )
-      .subscribe(roleResults => {
-        this.roles = roleResults;
-      });
-  }
+    const selectedRoles = this.roles.filter(r => r.selected);
 
-  addMemberWithRoles() {
-    if (this.selectedPrincipal.principalType === 'user') {
-      const user = new User(
-        this.configService.identityProvider,
-        this.selectedPrincipal.subjectId
-      );
+    if (selected.principalType === 'user') {
+      const user: IUser = { identityProvider: this.configService.identityProvider, subjectId: selected.subjectId };
       return this.userService
         .getUser(
           this.configService.identityProvider,
-          this.selectedPrincipal.subjectId
+          selected.subjectId
         )
-        .toPromise()
-        .then((userResult: User) => {
-          return userResult;
+        .mergeMap((userResult: IUser) => {
+          return Observable.of(userResult);
         })
         .catch(err => {
           if (err.statusCode === 404) {
-            return this.userService.createUser(user).toPromise();
+            return this
+              .userService
+              .createUser(user);
+          } else {
+            // TODO: handle error
           }
         })
-        .then((newUser: User) => {
+        .mergeMap((newUser: IUser) => {
           return this.userService
             .addRolesToUser(
               newUser.identityProvider,
               newUser.subjectId,
-              this.selectedRoles
-            )
-            .toPromise();
+              selectedRoles
+            );
+        })
+        .subscribe(() => {
+          this.router.navigate(['/accesscontrol']);
         });
     } else {
-      const group = new Group(this.selectedPrincipal.subjectId, '');
+      const group: IGroup = { groupName: selected.subjectId, groupSource: '' };
       return this.groupService
         .getGroup(group.groupName)
-        .toPromise()
-        .then((groupResult: Group) => {
-          return groupResult;
+        .mergeMap((groupResult: IGroup) => {
+          return Observable.of(groupResult);
         })
         .catch(err => {
           if (err.statusCode === 404) {
-            return this.groupService.createGroup(group).toPromise();
+            return this
+              .groupService
+              .createGroup(group);
+          } else {
+            // TODO: handle error
           }
         })
-        .then((newGroup: Group) => {
-          this.groupService
-            .addRolesToGroup(newGroup.groupName, this.selectedRoles)
-            .toPromise();
+        .mergeMap((newGroup: IGroup) => {
```

Comment:
Off topic, but did you see Ben Lesh's announcement that [`mergeMap` is changing to `smooshMap`?](https://twitter.com/benlesh/status/980336499970342913)